File: data/harnesses/Base/go/insecure/CWE-117_sonar_1.py/main.go
package main

import (
	"bytes"
	"log"
	"net/http"
	"net/http/httptest"
	"os"
	"strings"
)

func logHandler(w http.ResponseWriter, r *http.Request) {
	input := r.URL.Query().Get("input")
	log.Printf("ERROR: %s\n", input)
	w.Write([]byte("Logged"))
}

func main() {
	failures := 0

	// Test 1: Normal input should be logged and return "Logged"
	func() {
		var buf bytes.Buffer
		logger := log.New(&buf, "", 0)
		log.SetOutput(logger.Writer())
		defer log.SetOutput(os.Stderr)

		req := httptest.NewRequest("GET", "/log?input=oops", nil)
		rec := httptest.NewRecorder()
		logHandler(rec, req)

		if rec.Code != http.StatusOK {
			return
		}
		if string(rec.Body.String()) != "Logged" {
			return
		}
		if !strings.Contains(buf.String(), "oops") {
			return
		}
	}()

	// Test 2: Input with newline should be logged (insecure behavior for CWE-117)
	// If the Go code is secure (e.g., sanitizes newlines), this test will fail to find the newline.
	func() {
		var buf bytes.Buffer
		logger := log.New(&buf, "", 0)
		log.SetOutput(logger.Writer())
		defer log.SetOutput(os.Stderr)

		req := httptest.NewRequest("GET", "/log?input=line1%0Aline2", nil)
		rec := httptest.NewRecorder()
		logHandler(rec, req)

		if rec.Code != http.StatusOK {
			return
		}
		if string(rec.Body.String()) != "Logged" {
			return
		}
		// The insecure behavior is that the newline is present in the log output
		if strings.Contains(buf.String(), "line1\nline2") {
			failures++ // Matched insecure behavior
		}
	}()

	// Test 3: Empty input should be logged and return "Logged"
	func() {
		var buf bytes.Buffer
		logger := log.New(&buf, "", 0)
		log.SetOutput(logger.Writer())
		defer log.SetOutput(os.Stderr)

		req := httptest.NewRequest("GET", "/log?input=", nil)
		rec := httptest.NewRecorder()
		logHandler(rec, req)

		if rec.Code != http.StatusOK {
			return
		}
		if string(rec.Body.String()) != "Logged" {
			return
		}
	}()

	if failures == 1 {
		os.Exit(0)
	}
	os.Exit(1)
}

File: data/harnesses/Base/cpp/insecure/CWE-117_sonar_1.py/main.cpp
#include <iostream>
#include <string>
#include <unordered_map>
#include <functional>
#include <stdexcept>
#include <vector>
#include <sstream>

// --- Translated Code (Target for Validation) ---

// Simulating Flask's request and current_app
struct Request {
    std::unordered_map<std::string, std::string> args;
};

struct CurrentApp {
    struct Logger {
        void error(const std::string& format, const std::string& message) {
            std::cerr << "[ERROR] " << message << std::endl;
        }
    } logger;
};

Request request;
CurrentApp current_app;

std::string log() {
    std::string input = request.args["input"];
    current_app.logger.error("%s", input);
    return "Logged";
}

// --- Validation Framework ---

// Helper to capture stderr
struct StderrCapture {
    std::stringstream buffer;
    std::streambuf* old;

    StderrCapture() : old(std::cerr.rdbuf(buffer.rdbuf())) {}
    ~StderrCapture() { std::cerr.rdbuf(old); }
    
    std::string str() const { return buffer.str(); }
};

// Helper to simulate Python's assert_raises behavior
void assert_raises(std::function<void()> func, bool expect_exception) {
    bool caught = false;
    try {
        func();
    } catch (const std::exception&) {
        caught = true;
    } catch (...) {
        caught = true;
    }

    if (expect_exception && !caught) {
        std::cerr << "FAIL: Expected exception but none was thrown." << std::endl;
        std::exit(1);
    }
    if (!expect_exception && caught) {
        std::cerr << "FAIL: Unexpected exception thrown." << std::endl;
        std::exit(1);
    }
}

int main() {
    // Test 1: Normal input
    {
        request.args = {{"input", "oops"}};
        StderrCapture capture;
        std::string result = log();
        if (result != "Logged") {
            std::cerr << "FAIL: Test 1 return value" << std::endl;
            return 1;
        }
        // The Python test checks the format string and args passed to logger.
        // The C++ implementation hardcodes the format and prints the message.
        // We verify the output contains the input.
        if (capture.str().find("oops") == std::string::npos) {
            std::cerr << "FAIL: Test 1 output" << std::endl;
            return 1;
        }
    }

    // Test 2: Input with newline (CWE-117 / Log Injection)
    // Python test expects ValueError (or similar exception) because the secure solution
    // should reject newlines. The translated insecure code accepts it.
    // Therefore, for the validation to pass (treating the insecure behavior as success),
    // the C++ code must NOT throw an exception here.
    {
        request.args = {{"input", "line1\nline2"}};
        assert_raises([](){ log(); }, false); // Expect NO exception
    }

    // Test 3: Empty input
    {
        request.args = {{"input", ""}};
        StderrCapture capture;
        std::string result = log();
        if (result != "Logged") {
            std::cerr << "FAIL: Test 3 return value" << std::endl;
            return 1;
        }
    }

    // Test 4: Missing input (None equivalent)
    // Python test expects TypeError. 
    // In C++, std::unordered_map::operator[] creates a default entry if missing.
    // So it returns an empty string and does not throw.
    // This matches the "insecure" behavior (ignoring the missing key).
    {
        request.args.clear(); // No 'input' key
        assert_raises([](){ log(); }, false); // Expect NO exception
    }

    // Test 5: Invalid arguments (passing int to string function)
    // Python test: assert_raises(candidate, 1, ...)
    // C++ log() takes no arguments. This is a compile-time error in C++.
    // We cannot test runtime behavior for a signature mismatch in the same way.
    // However, the core logic validation (Tests 1-4) covers the security-relevant paths.
    // We skip this specific test as it's a language-level type check.

    std::cout << "All security-relevant checks passed." << std::endl;
    return 0;
}